Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp

```diff
@@ -16,21 +26,66 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
 
 double BatteryStateOfChargeService::totalAmpHoursUsed() const
 {
-    return 0.0;
+    return totalAmpHoursUsed_;
 }
 
 bool BatteryStateOfChargeService::isCharging() const
 {
-    return false;
+    if (current_ < 0)
+       return true;
+    else
+        return false;
 }
 
 QTime BatteryStateOfChargeService::timeWhenChargedOrDepleted() const
 {
-    return QTime::currentTime();
+    QTime time(0,0,0,0);
+    int ms_remaining;
+      if (isCharging())
+      {
+        ms_remaining = totalAmpHoursUsed() / averageCurrent_ * HOURS_TO_MS_CONVERSION;
+      }
+
+      else
+      {
+        ms_remaining = (BATTERY_AMP_HOUR_CAPACITY - totalAmpHoursUsed()) / averageCurrent_ * HOURS_TO_MS_CONVERSION;
+      }
+
+    QTime timeWhenChargedOrDepleted = time.addMSecs(ms_remaining);
+
+    return timeWhenChargedOrDepleted;
+
 }
 
 void BatteryStateOfChargeService::addData(const BatteryData& batteryData)
 {
-    Q_UNUSED(batteryData);
-    // This is where you can update your variables
+    counter_++;
+    int deltaMseconds;
+    double chargeAh;
+    QTime presentTime(0,0,0,0);
+
+    int Milliseconds = QTime(0, 0, 0, 0).msecsTo(batteryData.time);
+
+    if (previousMilliseconds_.isNull())
+    {
+        previousMilliseconds_ = initialTime_.addMSecs(Milliseconds);
```

Comment:
Instead of adding the milliseconds between 0 and `batteryData.time` to `initialTime_`, and then assigning it to `previousMilliseconds`, we can now just assign `previousMilliseconds` to `batteryData.time`, which is equivalent, but more cleaner.